include/vcpkg/base/optional.h

```diff
@@ -249,10 +249,22 @@ namespace vcpkg
         {
             constexpr OptionalStorage() noexcept : m_t(nullptr) { }
             constexpr OptionalStorage(const T& t) : m_t(&t) { }
-            constexpr OptionalStorage(const Optional<T>& t) : m_t(t.get()) { }
-            constexpr OptionalStorage(const Optional<const T>& t) : m_t(t.get()) { }
-            constexpr OptionalStorage(Optional<T>&& t) = delete;
-            constexpr OptionalStorage(Optional<const T>&& t) = delete;
+            template<typename U,
```

Comment:
Can you explain why this change was necessary? Optional stores T, not T*, so whether T is abstract should not matter. (`is_abstract` is almost never the right trait; usually people are looking for `is_constructible` or similar....)

Reply:
From Discord:

the problem is `OptionalStorage<const U&, B>` wanting a `Optional<U>` and `U = Interface` due to deduction
(https://godbolt.org/z/4Gf4e7PKT uncomment line 295 for the error)

Reply:
So the question is if the following is a valid expressions: `T a;` independent of the question if it can be constructed or not.

Reply:
Oh, I see, the problem is that given `Optional<const T&>` we try to form `Optional<T>` due to:

        template<class T, bool B>
        struct OptionalStorage<const T&, B>
        {
            // T doesn't have the const& here
            constexpr OptionalStorage(const Optional<T>& t) : m_t(t.get()) { }
        };

I believe the correct fix is just to put the `const&` back on rather than this `SFINAE`. That is, in `struct OptionalStorage<T&, B>`,

    constexpr OptionalStorage(Optional<T>& t) : m_t(t.get()) { }

should be

    constexpr OptionalStorage(Optional<T&>& t) : m_t(t.get()) { }

and in `struct OptionalStorage<const T&, B>`,

            constexpr OptionalStorage(const Optional<T>& t) : m_t(t.get()) { }
            constexpr OptionalStorage(const Optional<const T>& t) : m_t(t.get()) { }
            constexpr OptionalStorage(Optional<T>&& t) = delete;
            constexpr OptionalStorage(Optional<const T>&& t) = delete;

should be

            constexpr OptionalStorage(const Optional<const T&>& t) : m_t(t.get()) { }
            constexpr OptionalStorage(Optional<const T&>&& t) = delete;